File: backend/app/schemas/map.py
from datetime import date

from pydantic import BaseModel


class MapPoint(BaseModel):
    lat: float
    lng: float


class SubstationMarker(BaseModel):
    ss_code: int
    ss_name: str | None
    ss_type: str | None
    volt_class: str | None
    no_of_ptrs: int | None
    ss_doc: str | None
    lat: float
    lng: float


class SubstationLookup(BaseModel):
    ss_code: int
    title: str
    lat: float
    lng: float


class CountByCategory(BaseModel):
    category: str
    count: int


class SolarPlantMarker(BaseModel):
    solar_id: int
    plant_name: str | None
    installed_capacity_mw: float | None
    interfacing_ss: str | None
    lat: float
    lng: float


class EhvConsumerMarker(BaseModel):
    ehv_id: int
    name: str | None
    installed_capacity_mw: float | None
    substation: str | None
    feeder_id: int | None
    lat: float
    lng: float


class LineFeature(BaseModel):
    feeder_id: int
    feeder_name: str | None
    volt_class: str | None
    from_substation: str | None
    to_substation: str | None
    tower_count: int | None
    length_ckm: float | None
    path: list[MapPoint]


class TowerMarker(BaseModel):
    """Carries everything the legacy tower circle showed.

    arcgisScript.js bound a hover label of LOCATION NO + TYPE OF TOWER, and a
    click popup that mixed tower fields with line fields (feeder name, length,
    circuit/conductor type, date of charging). The line fields are joined in
    here rather than looked up client-side, because the viewport query returns
    towers spanning many feeders - the caller has no guarantee of holding the
    matching line.
    """

    tower_id: int
    feeder_id: int | None
    seq_no: int | None
    location_no: str | None
    tower_type: str | None
    # drives the legacy colour rules: yellow when a joint box is present,
    # orange when additional_info is "UC"
    telecom_joint_box: str | None
    additional_info: str | None
    lat: float
    lng: float

    # joined from gis.line
    line_volt_class: str | None
    line_feeder_name: str | None
    line_length_ckm: float | None
    line_tower_count: int | None
    line_circuit_type: str | None
    line_conductor_type: str | None
    line_date_of_charging: date | None


class PgcilSubstationMarker(BaseModel):
    id: int
    voltage: str | None
    name: str | None
    lat: float
    lng: float


class HydelPowerStationMarker(BaseModel):
    hydel_id: int
    name: str | None
    gen_cap_mw: float | None
    connected_ss: str | None
    lat: float
    lng: float


class PgcilLineMarker(BaseModel):
    id: int
    feeder_name: str | None
    lat: float
    lng: float

File: web/src/types/map.ts
// Mirrors app/schemas/map.py exactly - field-for-field, so a backend schema
// change is a compile error here rather than a silent runtime mismatch.

export interface MapPoint {
  lat: number;
  lng: number;
}

export interface SubstationMarker {
  ss_code: number;
  ss_name: string | null;
  ss_type: string | null;
  volt_class: string | null;
  no_of_ptrs: number | null;
  ss_doc: string | null;
  primary_mva_cap: number | null;
  district: string | null;
  zone: string | null;
  circle: string | null;
  division: string | null;
  /** Filenames only - joined to VITE_DOCUMENT_BASE_URL by the client. */
  link_sld: string | null;
  link_ss_photo: string | null;
  lat: number;
  lng: number;
}

export interface SubstationLookup {
  ss_code: number;
  title: string;
  lat: number;
  lng: number;
}

export interface CountByCategory {
  category: string;
  count: number;
}

export interface SolarPlantMarker {
  solar_id: number;
  plant_name: string | null;
  installed_capacity_mw: number | null;
  interfacing_ss: string | null;
  lat: number;
  lng: number;
}

export interface EhvConsumerMarker {
  ehv_id: number;
  name: string | null;
  installed_capacity_mw: number | null;
  substation: string | null;
  feeder_id: number | null;
  lat: number;
  lng: number;
}

export interface LineFeature {
  feeder_id: number;
  feeder_name: string | null;
  volt_class: string | null;
  from_substation: string | null;
  to_substation: string | null;
  tower_count: number | null;
  length_ckm: number | null;
  path: MapPoint[];
}

export interface TowerMarker {
  tower_id: number;
  feeder_id: number | null;
  seq_no: number | null;
  location_no: string | null;
  tower_type: string | null;
  // drive the legacy colour rules
  telecom_joint_box: string | null;
  additional_info: string | null;
  lat: number;
  lng: number;

  // joined from gis.line, because a viewport query spans many feeders
  line_volt_class: string | null;
  line_feeder_name: string | null;
  line_length_ckm: number | null;
  line_tower_count: number | null;
  line_circuit_type: string | null;
  line_conductor_type: string | null;
  line_date_of_charging: string | null;
}

export interface PgcilSubstationMarker {
  id: number;
  voltage: string | null;
  name: string | null;
  lat: number;
  lng: number;
}

export interface HydelPowerStationMarker {
  hydel_id: number;
  name: string | null;
  gen_cap_mw: number | null;
  connected_ss: string | null;
  lat: number;
  lng: number;
}

export interface PgcilLineMarker {
  id: number;
  feeder_name: string | null;
  lat: number;
  lng: number;
}
